\documentclass{article}

%\usepackage[a4paper, top=3cm,left=0.5cm,right=0.5cm,bottom=2cm,headheight=22.5pt]{geometry}
\usepackage[a4paper, left=0.5cm,right=0.5cm,bottom=2cm, headheight=4cm]{geometry}
\usepackage{tcolorbox}
\usepackage{fancyhdr}
\usepackage{multirow}
\usepackage{collcell}
\usepackage{colortbl}
\usepackage{array}
\usepackage{tabu}
\usepackage[capposition=top]{floatrow}
%\usepackage[table]{xcolor}
\pagestyle{fancy}

%\newcommand*\approval[2][5cm]{\vspace*{2cm}\parbox{#1}{\hrulefill\par#2}}

%\lhead{\includegraphics[width=4cm, height=1.3cm]{data/logo}}
%\lhead{\includegraphics[width=\linewidth]{data/logo}}

%\fancyhead[L]{\includegraphics[scale=0.15]{data/logo}}
%\fancyhead[L]{\includegraphics[width=22pt]{data/logo}}
%\fancyhead[C]{\textbf{TUBERCULOSIS RESISTANCE REPORT} \\ \textit{Whole Genome Sequencing Report}}
%\fancyhead[R]{\textbf{PATIENT NHI:} \Sexpr{repData$NHI}}
%\fancyhf{}
%\lhead{\fbox{\parbox[b][4cm][c]{10cm}{\includegraphics[height=3.5cm]{data/logo}}}}
%\chead{\fbox{\parbox[b][4cm][c]{10cm}{TB REPORT \\ WGS}}}
%\rhead{\fbox{\parbox[b][4cm][c]{10cm}{ \today \\ \\ TOP \\ CONFIDENTIAL}}}

%\lhead{{\includegraphics[height=3.5cm]{data/logo}}
%\lhead{\makebox[\headwidth][c]{\includegraphics{data/logo}}}

%\fancyfoot[L]{PATIENT NHI: \Sexpr{repData$NHI}}
%\fancyfoot[R]{\today}

\fancyhf{}

\fancypagestyle{body}{
  \fancyhead[L]{\parbox[b][2cm][c]{10cm}{\includegraphics[height=3.5cm]{data/logo}}}
  \fancyhead[C]{\textbf{TUBERCULOSIS RESISTANCE REPORT} \\ \textit{Whole Genome Sequencing Report}}
  \fancyhead[R]{\textbf{PATIENT NHI:} \Sexpr{repData$NHI}}
  \fancyfoot[L]{PATIENT NHI: \Sexpr{repData$NHI}}
  \fancyfoot[R]{\today}
}


\fancypagestyle{ending}{
  \fancyhead[L]{\parbox[b][2cm][c]{10cm}{\includegraphics[height=3.5cm]{data/logo}}}
  \fancyhead[C]{\textbf{TUBERCULOSIS RESISTANCE REPORT} \\ \textit{Whole Genome Sequencing Report}}
  \fancyhead[R]{LAST PAGE}

  \fancyfoot[L]{\hspace{1cm}\textbf{Clinician:} \rule{5cm}{.4pt}}
  \fancyfoot[C]{\hspace{5cm}\textbf{Signature:} \rule{5cm}{.4pt}}
}


\newtcolorbox{reportSection}[2][]{colbacktitle=black!8!white, colback=white!5!white, coltitle=black!70!black, boxrule=0.5pt, title=\textbf{\large{#2}}, fonttitle=\bfseries,#1}


\newtcolorbox{reportBox}[2][]{colbacktitle=black!15!white, colback=white!5!white, coltitle=black!70!black, title=\textbf{\large{#2}}, fonttitle=\bfseries,#1}


\newcolumntype{U}{>{\collectcell\MakeUpperCase}l<{\endcollectcell}}
%\renewcommand{\arraystretch}{1.2}
%\renewcommand{\headrulewidth}{0pt}

\pagestyle{body}
\begin{document}

\SweaveOpts{concordance=TRUE}

%%% SECTION 1: Summary Report
\begin{reportSection}{A1: Summary Report}

%% PATIENT DETAILS
\begin{reportBox}{Patient Details}
\noindent \begin{tabular}{|p{0.16\textwidth}p{0.3\textwidth}||p{0.16\textwidth}p{0.3\textwidth}|}
\hline

Patient Name: & \uppercase{\Sexpr{repData$FName}} \uppercase{\Sexpr{repData$LName}} & Barcode: & \uppercase{\Sexpr{repData$BARCODE}} \\
DOB: & \Sexpr{repData$DOB} & NHI: & \uppercase{\Sexpr{repData$NHI}} \\
Gender: & \uppercase{\Sexpr{repData$GENDER}} & Ethnicity: & \uppercase{\Sexpr{repData$ETHNICITY}} \\ \cline{1-4}
\hline
Location: & \uppercase{\Sexpr{repData$Location}} & Sample Type: & \uppercase{\Sexpr{repData$sType}} \\
Sample Date: & \Sexpr{repData$sDate} & Culture Method: & \uppercase{\Sexpr{repData$sSeq}} \\
Reporting Lab: & \uppercase{\Sexpr{repData$reportLab}} & Technician: & DR ROD JACKSON \\
Sequencing Date: & \uppercase{\Sexpr{repData$reportDate}} & Sequencing: & WGS \\
\hline
Requested by: & \uppercase{\Sexpr{repData$requestor}} & Contact: & \uppercase{\Sexpr{repData$requestorContact}} \\ \cline{1-4}
\end{tabular}
\end{reportBox}

\vspace{1cm}

%% RESISTANCE SUMMARY
%\begin{reportBox}{Resistance Profile}
%\noindent \begin{tabular}{|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|} \hline
%\multicolumn{1}{|l|}{\textbf{Drug Class}} & \textbf{Resistance} & \textbf{Drug} & \textbf{Mutation} \\ \cline{1-4}

% <sec2:drugFunction,echo=FALSE,results=tex>>=

%# First Line (Sensitive and Resistance)
%FLSens <- filter(repDrug, CLASS == "First Line" ) %>% filter(STATUS == "Sensitive")
%FLRes <- filter(repDrug, CLASS == "First Line" ) %>% filter(STATUS == "Resistance")

%FLTotal <- nrow(FLSens) + nrow(FLRes)

%# Scripting in First Line Sensitive

%if(nrow(FLSens) > 0){
%  for( i in 1:nrow(FLSens) ){
%    if(i == nrow(FLSens)){
%      if(nrow(FLRes)==0){
%        cat(sprintf("\\multirow{-%d}{*}{First Line}", nrow(FLSens)), "&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(FLSens)), "&", FLSens[i,]$DRUG, "&", "\\textit{",FLSens[i,]$GENE,"}", "\\\\ \\cline{1-4}")
%      }
%      else{
%        cat("&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(FLSens)), "&", FLSens[i,]$DRUG, "&", "\\textit{",FLSens[i,]$GENE,"}", "\\\\ \\cline{2-4}")
%      }
%    }
%%      else{
%        cat("&&", FLSens[i,]$DRUG, "&", "\\textit{",FLSens[i,]$GENE,"}", "\\\\ \\cline{3-4}")
%      }
%    }
%  }


%if(nrow(FLRes) > 0){
%  for(i in 1:nrow(FLRes)){
%    if(i == nrow(FLRes)){
%      cat(sprintf("\\multirow{-%d}{*}{First Line}", FLTotal), "&", sprintf("\\textbf{\\multirow{-%d}{*}{\\cellcolor[gray]{.9}Resistance}}", nrow(FLRes)), "&", "\\cellcolor[gray]{.9} \\textbf{", FLRes[i,]$DRUG,"}", %"&", "\\cellcolor[gray]{.9} \\textbf{",FLRes[i,]$GENE,"}", "\\\\ \\cline{1-4}")
%    }
%    else{
%      cat("& \\cellcolor[gray]{.9} &", "\\cellcolor[gray]{.9} \\textbf{",FLRes[i,]$DRUG,"}", "&", "\\cellcolor[gray]{.9} \\textbf{",FLRes[i,]$GENE,"}", "\\\\ \\cline{3-4}")
%    }
%  }
%}


%\end{tabular}

%\vspace{1cm}

%\noindent \begin{tabular}{|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|} \hline
%\multicolumn{1}{|l|}{\textbf{Drug Class}} & \textbf{Resistance} & \textbf{Drug} & \textbf{Mutation} \\ \cline{1-4}

% < < sec3:drugFunction,echo=FALSE,results=tex>>=
%###########

%# Second Line (Sensitive and Resistance)
%SLSens <- filter(repDrug, CLASS == "Second Line" ) %>% filter(STATUS == "Sensitive")
%SLRes <- filter(repDrug, CLASS == "Second Line" ) %>% filter(STATUS == "Resistance")

%SLTotal <- nrow(SLSens) + nrow(SLRes)

%if(nrow(SLSens) > 0){
%  for( i in 1:nrow(SLSens) ){
%    if(i == nrow(SLSens)){
%      if(nrow(SLRes)==0){
%        cat(sprintf("\\multirow{-%d}{*}{Second Line}", nrow(SLSens)), "&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(SLSens)), "&", SLSens[i,]$DRUG, "&", "\\textit{",SLSens[i,]$GENE,"}", "\\\\ \\cline{1-4}")
%      }
%      else{
%        cat("&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(SLSens)), "&",SLSens[i,]$DRUG, "&", "\\textit{",SLSens[i,]$GENE,"}", "\\\\ \\cline{2-4}")
%      }
%    }
%      else{
%        cat("&&", SLSens[i,]$DRUG, "&", "\\textit{",SLSens[i,]$GENE,"}", "\\\\ \\cline{3-4}")
%      }
%    }
%  }


%if(nrow(SLRes) > 0){
%  for(i in 1:nrow(SLRes)){
%    if(i == nrow(SLRes)){
%      cat(sprintf("\\multirow{-%d}{*}{Second Line}", SLTotal), "&", sprintf("\\textbf{\\multirow{-%d}{*}{\\cellcolor[gray]{.9}Resistance}}", nrow(SLRes)), "&","\\cellcolor[gray]{.9} \\textbf{", SLRes[i,]$DRUG,"}", %"&", "\\cellcolor[gray]{.9} \\textbf{",SLRes[i,]$GENE,"}", "\\\\ \\cline{1-4}")
%    }
%    else{
%      cat("& \\cellcolor[gray]{.9} &", "\\cellcolor[gray]{.9} \\textbf{",SLRes[i,]$DRUG,"}", "&", "\\cellcolor[gray]{.9} \\textbf{",SLRes[i,]$GENE,"}", "\\\\ \\cline{3-4}")
  %  }
 % }
%}


%\end{tabular}
%\end{reportBox}

\begin{reportBox}{Resistance Profile}
\noindent \begin{tabular}{|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|} \hline
\multicolumn{1}{|l|}{\textbf{Drug Class}} & \textbf{Resistance} & \textbf{Drug} & \textbf{Mutation} \\ \cline{1-4}

<<sec4:drugFunction,echo=FALSE,results=tex>>=

# First Line (Sensitive and Resistance)
FLSens <- filter(repDrug, CLASS == "First Line" ) %>% filter(STATUS == "Sensitive")
FLRes <- filter(repDrug, CLASS == "First Line" ) %>% filter(STATUS == "Resistance")

FLTotal <- nrow(FLSens) + nrow(FLRes)

# Scripting in First Line Sensitive

if(nrow(FLSens) > 0){
  for( i in 1:nrow(FLSens) ){
    if(i == nrow(FLSens)){
      if(nrow(FLRes)==0){
        cat(sprintf("\\multirow{-%d}{*}{First Line}", nrow(FLSens)), "&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(FLSens)), "&", FLSens[i,]$DRUG, "&", "\\textit{",FLSens[i,]$GENE,"}", "\\\\ \\cline{1-4}")
      }
      else{
        cat("&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(FLSens)), "&", FLSens[i,]$DRUG, "&", "\\textit{",FLSens[i,]$GENE,"}", "\\\\ \\cline{2-4}")
      }
    }
      else{
        cat("&&", FLSens[i,]$DRUG, "&", "\\textit{",FLSens[i,]$GENE,"}", "\\\\ \\cline{3-4}")
      }
    }
  }


if(nrow(FLRes) > 0){
  for(i in 1:nrow(FLRes)){
    if(i == nrow(FLRes)){
      cat(sprintf("\\multirow{-%d}{*}{First Line}", FLTotal), "&", sprintf("\\textbf{\\multirow{-%d}{*}{\\cellcolor[gray]{.9}Resistance}}", nrow(FLRes)), "&", "\\cellcolor[gray]{.9} \\textbf{", FLRes[i,]$DRUG,"}", "&", "\\cellcolor[gray]{.9} \\textbf{",FLRes[i,]$GENE,"}", "\\\\ \\cline{1-4}")
    }
    else{
      cat("& \\cellcolor[gray]{.9} &", "\\cellcolor[gray]{.9} \\textbf{",FLRes[i,]$DRUG,"}", "&", "\\cellcolor[gray]{.9} \\textbf{",FLRes[i,]$GENE,"}", "\\\\ \\cline{3-4}")
    }
  }
}

# Second Line (Sensitive and Resistance)
SLSens <- filter(repDrug, CLASS == "Second Line" ) %>% filter(STATUS == "Sensitive")
SLRes <- filter(repDrug, CLASS == "Second Line" ) %>% filter(STATUS == "Resistance")

SLTotal <- nrow(SLSens) + nrow(SLRes)

if(nrow(SLSens) > 0){
  for( i in 1:nrow(SLSens) ){
    if(i == nrow(SLSens)){
      if(nrow(SLRes)==0){
        cat(sprintf("\\multirow{-%d}{*}{Second Line}", nrow(SLSens)), "&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(SLSens)), "&", SLSens[i,]$DRUG, "&", "\\textit{",SLSens[i,]$GENE,"}", "\\\\ \\cline{1-4}")
      }
      else{
        cat("&", sprintf("\\multirow{-%d}{*}{Sensitive}", nrow(SLSens)), "&",SLSens[i,]$DRUG, "&", "\\textit{",SLSens[i,]$GENE,"}", "\\\\ \\cline{2-4}")
      }
    }
      else{
        cat("&&", SLSens[i,]$DRUG, "&", "\\textit{",SLSens[i,]$GENE,"}", "\\\\ \\cline{3-4}")
      }
    }
  }


if(nrow(SLRes) > 0){
  for(i in 1:nrow(SLRes)){
    if(i == nrow(SLRes)){
      cat(sprintf("\\multirow{-%d}{*}{Second Line}", SLTotal), "&", sprintf("\\textbf{\\multirow{-%d}{*}{\\cellcolor[gray]{.9}Resistance}}", nrow(SLRes)), "&","\\cellcolor[gray]{.9} \\textbf{", SLRes[i,]$DRUG,"}", "&", "\\cellcolor[gray]{.9} \\textbf{",SLRes[i,]$GENE,"}", "\\\\ \\cline{1-4}")
    }
    else{
      cat("& \\cellcolor[gray]{.9} &", "\\cellcolor[gray]{.9} \\textbf{",SLRes[i,]$DRUG,"}", "&", "\\cellcolor[gray]{.9} \\textbf{",SLRes[i,]$GENE,"}", "\\\\ \\cline{3-4}")
    }
  }
}
@
\end{tabular}
\end{reportBox}
\end{reportSection}

\pagebreak

%%% SECTION 2: WGS Technical Report

%% ALL DETECTED MUTATIONS
\begin{reportSection}{A2: Whole Genome Sequencing Report}

\begin{reportBox}{Genotypic Report}
\begin{tabular}{|l|l|l|l|l|l|l|}
\hline
\multirow{2}{*}{Drug} & Genomics  & \multicolumn{2}{c}{Allele} & Gene & \multirow{2}{*}{High Confidence}  & \multirow{2}{*}{PMID}\\ \cline{3-4}
& Position & Reference & Alternative & (Mutation) & & \\ \cline{1-7}

<<sec4:snpRESULTS,echo=FALSE,results=tex>>=

drugs <- c("Isoniazid", "Rifampicin", "Ethambutol", "Pyrazinamide", "Amikacin", "Capreomycin", "Ethionamide", "Kanamycin", "Streptomycin", "Linezolid", "Para-aminosalicylic", "Fluoroquinolones")

for( i in drugs ){
  snps <- filter(repSNP, DRUG == i)

  if( nrow(snps) > 0){
    for ( i in 1:nrow(snps)){
          cat(snps[i,]$DRUG, "&", snps[i,]$POSITION, "&",snps[i,]$REFERENCE, "&",snps[i,]$ALTERNATE, "&",snps[i,]$GENE, "&",snps[i,]$CONFIDENCE, "&",snps[i,]$PMID, "\\\\ \\cline{1-7}")
    }
  }
}

@


\end{tabular}
\vspace{0.7cm} \\
\parbox{16.5cm}{\small{\textbf{NOTE:} This genotypic report is a prediction of drug resistance based on the detection of known resistance conferring mutations, and should only be used as a guide. The actual drug susceptability may differ from what is stated in this report.}}


\end{reportBox}

\end{reportSection}
\pagebreak

% Modify the footer to include the approval section
\pagestyle{ending}

\begin{reportSection}{A3: Technical Report}

\begin{reportBox}{Pipeline}
\begin{tabular}{|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|}
\hline
Pipeline & \Sexpr{repMeta$PIPELINE} & Genome & \Sexpr{repMeta$GENOME} \\
Phred Score & 30 & Masking & \Sexpr{repMeta$MASKING} \\ \cline{1-4}
\end{tabular}
\end{reportBox}



\begin{reportBox}{Forward Statistics: WRX.R1.fastq.gz}
\vspace{0.5cm}
\begin{tabular}{|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|}
\hline
\multicolumn{2}{l}{Pre-Trimming} & \multicolumn{2}{l}{Post-Trimming} \\
Total Reads & VALUE & Total Reads & VALUE (\%) \\
Average Quality & VALUE & Average Quality & VALUE \\
GC\% & VALUE & GC\% & VALUE \\
Mean read length & VALUE & Mean read length & VALUE \\
Read range & VALUE & Read range & VALUE \\ \cline{1-4}
\end{tabular}
\end{reportBox}

\begin{reportBox}{Reverse Statistics}
\begin{tabular}{|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|p{0.2\textwidth}|}
\hline
File name & NAME && \\
\hline
\multicolumn{2}{c}{Pre-Trimming} & \multicolumn{2}{c}{Post-Trimming} \\
Total Reads & VALUE & Total Reads & VALUE (\%) \\
Average Quality & VALUE & Average Quality & VALUE \\
GC\% & VALUE & GC\% & VALUE \\
Mean read length & VALUE & Mean read length & VALUE \\
Read range & VALUE & Read range & VALUE \\ \cline{1-4}
\end{tabular}
\end{reportBox}

\begin{reportBox}{Alignment Statistics}
\end{reportBox}

\begin{reportBox}{Notes}
\end{reportBox}

%\begin{tabular}{|p{3cm}|p{3cm}|p{3cm}|}
%\begin{tabular}{|p{0.3\textwidth}|p{0.3\textwidth}|p{0.3\textwidth}|}
%\hline
%Field 1 & Filed 2 & Filed 3
%\end{tabular}

\end{reportSection}
\end{document}